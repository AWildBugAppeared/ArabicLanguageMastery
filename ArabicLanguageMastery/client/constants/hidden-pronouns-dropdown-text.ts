export const HiddenPronounsDropdownText = {
  faailPresent: 'No Hidden Faail',
  huwa: 'هو',
  hiya: 'هي',
  anta: 'أنت',
  ana: 'أنا',
  nahnu: 'نحن',
};
