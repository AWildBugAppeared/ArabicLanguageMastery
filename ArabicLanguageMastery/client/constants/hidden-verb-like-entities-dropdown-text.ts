export const HiddenVerbLikeEntitiesDropdownText = {
  shibhulFilPresent: 'No Hidden VLE',
  mawjuwdun: 'مَوْجُوْدٌ',
  mawjuwdan: 'مَوْجُوْدًا',
  mawjuwdin: 'مَوْجُوْدٍ',
  mawjuwdatun: 'مَوْجُوْدَةٌ',
  mawjuwdatan: 'مَوْجُوْدَةً',
  mawjuwdatin: 'مَوْجُوْدَةٍ',
};
