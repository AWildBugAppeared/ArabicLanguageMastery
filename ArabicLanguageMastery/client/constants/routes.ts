const introductoryModuleRoute = 'introductory-module';

export const Routes = {
  introductoryModule: {
    theArabicSciences: `/${introductoryModuleRoute}/the-arabic-sciences`,
    introductionToSarf: `/${introductoryModuleRoute}/introduction-to-sarf`,
    introductionToNahwPart1: `/${introductoryModuleRoute}/introduction-to-nahw-part-1-the-two-types-of-sentences`,
    theMaadhiVerb: `/${introductoryModuleRoute}/the-maadhi-verb`,
    introductionToPhrases: `/${introductoryModuleRoute}/introduction-to-phrases`,
    theDescriptivePhrase: `/${introductoryModuleRoute}/the-descriptive-phrase`,
    thePossessivePhrase: `/${introductoryModuleRoute}/the-possessive-phrase`,
    theThreePartsOfSpeech: `/${introductoryModuleRoute}/the-three-parts-of-speech`,
  },
  contactUs: '/contact-us',
  englishToArabic: '/english-to-arabic',
};
