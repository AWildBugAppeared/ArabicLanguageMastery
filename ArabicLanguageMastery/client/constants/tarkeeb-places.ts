export const TarkeebPlaces = {
  ismInna: 'اسم إن',
  ismKaana: 'اسم كان',
  jaar: 'جارّ',
  haal: 'حال',
  khabr: 'خبر',
  khabrInna: 'خبر إن',
  khabrKaana: 'خبر كان',
  dhulHaal: 'ذو الحال',
  atf: 'عطف',
  faail: 'فاعل',
  fil: 'فعل',
  mafoolBiHi: 'مفعول به',
  mafoolFiyHi: 'مفعول فيه',
  mafoolLaHu: 'مفعول له',
  mafoolMutlaq: 'مفعول مطلق',
  majroor: 'مجرور',
  matoof: 'معطوف',
  matoofAlayh: 'معطوف عليه',
  mubtada: 'مبتدأ',
  mudaaf: 'مضاف',
  mudaafIlayhi: 'مضاف إليه',
  mutaalliq: 'متعلّق',
};
