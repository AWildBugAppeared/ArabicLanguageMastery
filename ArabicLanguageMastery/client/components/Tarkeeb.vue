<template>
  <v-container class="arabic-l text-center">
    <div ref="answer">
      <v-runtime-template
        class="answer"
        :template="answer"
      ></v-runtime-template>
    </div>

    <div dir="rtl">
      <v-btn
        v-for="tarkeebPlace in tarkeebPlaces"
        :id="tarkeebPlace"
        :key="tarkeebPlace"
        class="mx-2 px-2 pb-5 pt-6 primary"
        @click="setBox(tarkeebPlace)"
        ><span class="arabic">{{ tarkeebPlace }}</span></v-btn
      >
    </div>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue';

import VRuntimeTemplate from 'v-runtime-template';
import { TarkeebPlaces } from '~/constants/tarkeeb-places';

export default Vue.extend({
  name: 'Tarkeeb',
  components: { VRuntimeTemplate },

  props: {
    arabic: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      answer: '',
      arabicCharArray: [] as string[],
      firstSelectedIndex: -1,
      secondSelectedIndex: -1,
      tarkeebPlaces: Object.values(TarkeebPlaces),
    };
  },

  created() {
    this.arabicCharArray = this.arabic.split('');

    this.answer += '<div id="answer">';

    let isNonConsonantCharacter = false;

    for (let i = 0; i < this.arabicCharArray.length; i++) {
      isNonConsonantCharacter = this.isNonConsonantCharacter(
        this.arabicCharArray[i]
      );

      this.answer += `<span id="${i}"${
        isNonConsonantCharacter ? '' : ` @click="setSelected(${i})"`
      }>${
        this.arabicCharArray[i] === ' ' ? '&nbsp;' : this.arabicCharArray[i]
      }</span>`;
    }

    this.answer += '</div>';
  },

  methods: {
    isNonConsonantCharacter(character: string) {
      return [
        '\u064B',
        '\u064C',
        '\u064D',
        '\u064E',
        '\u064F',
        '\u0650',
        '\u0651',
        '\u0652',
        '\u0653',
        '\u0654',
        '\u0655',
        '\u0656',
        '\u0657',
        ' ',
      ].includes(character);
    },

    removeSelection(id: string) {
      this.answer = this.answer.replace(
        `<span id="${id}" class="selected" @click="setSelected(${id})"`,
        `<span id="${id}" @click="setSelected(${id})"`
      );
    },

    setSelected(id: string) {
      const spanId = parseInt(id);

      if (this.firstSelectedIndex === spanId) {
        this.firstSelectedIndex = -1;
        this.removeSelection(id);
        return;
      }

      if (this.secondSelectedIndex === spanId) {
        this.secondSelectedIndex = -1;
        this.removeSelection(id);
        return;
      }

      if (this.firstSelectedIndex > -1 && this.secondSelectedIndex > -1) {
        return;
      }

      if (this.firstSelectedIndex === -1) {
        this.firstSelectedIndex = parseInt(id);
      } else if (this.secondSelectedIndex === -1) {
        this.secondSelectedIndex = parseInt(id);

        // This is done to support the user highlight from rtl as well as ltr
        if (this.firstSelectedIndex > this.secondSelectedIndex) {
          const originalFirstSelectedIndex = this.firstSelectedIndex;
          this.firstSelectedIndex = this.secondSelectedIndex;
          this.secondSelectedIndex = originalFirstSelectedIndex;
        }
      }

      this.answer = this.answer.replace(
        `<span id="${id}"`,
        `<span id="${id}" class="selected"`
      );
    },

    removeBox(id: string) {
      const regexString = `<fieldset id="${id}">.+<!-- ${id.replace(
        'f',
        'l'
      )} --><\\/legend>`;
      const regex = new RegExp(regexString);
      this.answer = this.answer.replace(regex, '');
      this.answer = this.answer.replace(`<!-- ${id} --></fieldset>`, '');
    },

    setBox(tarkeebPlace: string) {
      if (this.firstSelectedIndex === -1 && this.secondSelectedIndex === -1) {
        return;
      }

      const randomValues = new Uint32Array(1);
      crypto.getRandomValues(randomValues);

      const id = randomValues[0];

      const fieldSetStartTag = `<fieldset id="${
        'fs-' + id
      }"><legend><span @click="removeBox('fs-' + ${id})">${tarkeebPlace}</span><!-- ${
        'ls-' + id
      } --></legend>`;
      const fieldSetEndTag = `<!-- ${'fs-' + id} --></fieldset>`;

      const regexString =
        this.firstSelectedIndex > -1 && this.secondSelectedIndex > -1
          ? `<span id="${this.firstSelectedIndex}" class="selected" @click="setSelected\\(${this.firstSelectedIndex}\\)">.+<span id="${this.secondSelectedIndex}" class="selected" @click="setSelected\\(${this.secondSelectedIndex}\\)">.<\\/span>`
          : `<span id="${this.firstSelectedIndex}" class="selected" @click="setSelected\\(${this.firstSelectedIndex}\\)">.<\\/span>`;

      const regex = new RegExp(regexString, 'g');

      const matches = this.answer.match(regex);

      if (!matches) {
        return;
      }

      let replaceAnswerString = '';

      replaceAnswerString += `${fieldSetStartTag}${matches[0]}`;

      const highestSelectedIndex =
        this.secondSelectedIndex > -1
          ? this.secondSelectedIndex
          : this.firstSelectedIndex;
      const nextCharacter = this.arabicCharArray[highestSelectedIndex + 1];

      if (
        nextCharacter &&
        this.isNonConsonantCharacter(nextCharacter) &&
        nextCharacter !== ' '
      ) {
        replaceAnswerString +=
          this.answer.match(
            new RegExp(`<span id="${highestSelectedIndex + 1}">.<\\/span>`)
          )?.[0] ?? '';

        this.answer = this.answer.replace(
          new RegExp(`<span id="${highestSelectedIndex + 1}">.<\\/span>`),
          ''
        );

        replaceAnswerString += fieldSetEndTag;
      } else {
        replaceAnswerString += fieldSetEndTag;
      }

      replaceAnswerString = replaceAnswerString.replaceAll(
        ' class="selected"',
        ''
      );

      this.answer = this.answer.replace(regex, replaceAnswerString);

      this.firstSelectedIndex = this.secondSelectedIndex = -1;
    },
  },
});
</script>

<style lang="scss">
$fieldset-color: purple;
$first-child-fieldset-color: blue;
$second-child-fieldset-color: green;
$third-child-fieldset-color: orange;

fieldset {
  border-color: $fieldset-color;
  display: inline-block;
  direction: rtl;
  padding: 0 8px 8px 8px;
  font-size: unset;
}

fieldset {
  legend {
    font-size: 0.8em;
    line-height: 1;
    color: $fieldset-color;
  }

  fieldset {
    border-color: $first-child-fieldset-color;
    legend {
      color: $first-child-fieldset-color;
    }

    fieldset {
      border-color: $second-child-fieldset-color;
      legend {
        color: $second-child-fieldset-color;
      }

      fieldset {
        border-color: $third-child-fieldset-color;
        legend {
          color: $third-child-fieldset-color;
        }
      }
    }
  }
}

.answer {
  margin-bottom: 30px;
  direction: rtl;
}

.selected {
  background-color: green;
}
</style>
