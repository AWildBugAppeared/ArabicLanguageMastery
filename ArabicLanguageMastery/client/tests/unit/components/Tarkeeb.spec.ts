/* eslint-disable  @typescript-eslint/no-explicit-any */

// Imports
import Vue from 'vue';
import Vuetify from 'vuetify';

// Utilities
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';

import Tarkeeb from '@/components/Tarkeeb.vue';
import { toMatchImageSnapshot } from 'jest-image-snapshot';
import { TarkeebPlaces } from '~/constants/tarkeeb-places';

const puppeteer = require('puppeteer');
expect.extend({ toMatchImageSnapshot });

describe('Tarkeeb.vue', () => {
  const localVue = createLocalVue();
  let wrapper: Wrapper<Vue & { [key: string]: any }>;
  let vuetify: Vuetify;

  const mountWrapper = () => {
    wrapper = mount(Tarkeeb, {
      localVue,
      vuetify,
      propsData: {
        arabic: 'وَ رَجَعَ النَّاسُ وَ دَخَلُوْا فِيْ بَيْتِ الْأَصْنَامِ',
      },
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify({
      mocks: {
        $vuetify: {},
      },
    });
    Vue.use(Vuetify);
    document.body.setAttribute('data-app', 'true');
  });

  afterEach(() => {
    wrapper.destroy();
    jest.resetModules();
    jest.clearAllMocks();
  });

  describe.only('render', () => {
    it('should render component with Arabic props passed in', async () => {
      mountWrapper();
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should add colour selection', async () => {
      const browser = await puppeteer.launch();
      const page = await browser.newPage();
      await page.goto('http://localhost:8000/tarkeeb-test');

      await page.click('span[id="10"]');
      const image = await page.screenshot();

      expect(image).toMatchImageSnapshot();
    });

    it('should remove colour selection if clicked and is already selected', async () => {
      const browser = await puppeteer.launch();
      const page = await browser.newPage();
      await page.goto('http://localhost:8000/tarkeeb-test');

      await page.click('span[id="10"]');
      await page.click('span[id="10"]');
      const image = await page.screenshot();

      expect(image).toMatchImageSnapshot();
    });

    it('should apply nested boxes with correct colours', async () => {
      const browser = await puppeteer.launch();
      const page = await browser.newPage();
      await page.goto('http://localhost:8000/tarkeeb-test');

      await page.click('span[id="29"]');
      await page.click('span[id="61"]');
      await page.click(`#${TarkeebPlaces.matoof}`);

      await page.click('span[id="29"]');
      await page.click('span[id="33"]');
      await page.click(`#${TarkeebPlaces.fil}`);

      await page.click('span[id="35"]');
      await page.click(`#${TarkeebPlaces.faail}`);

      await page.click('span[id="39"]');
      await page.click('span[id="61"]');
      await page.click(`#${TarkeebPlaces.mutaalliq}`);

      await page.click('span[id="39"]');
      await page.click('span[id="41"]');
      await page.click(`#${TarkeebPlaces.jaar}`);

      await page.click('span[id="44"]');
      await page.click('span[id="61"]');
      await page.click(`#${TarkeebPlaces.majroor}`);

      await page.click('span[id="44"]');
      await page.click('span[id="48"]');
      await page.click(`#${TarkeebPlaces.mudaaf}`);

      await page.click('span[id="51"]');
      await page.click('span[id="61"]');
      await page.click('#مضاف\\ إليه');

      await page.evaluate(() => {
        window.scroll(0, 0);
      });
      const image = await page.screenshot();

      expect(image).toMatchImageSnapshot();
    });
  });
});
