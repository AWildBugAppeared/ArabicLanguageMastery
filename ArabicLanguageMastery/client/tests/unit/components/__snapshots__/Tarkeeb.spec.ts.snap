// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tarkeeb.vue render should render component with Arabic props passed in 1`] = `
"<div class=\\"container arabic-l text-center\\">
  <div>
    <div id=\\"answer\\" class=\\"answer\\"><span id=\\"0\\">و</span><span id=\\"1\\">َ</span><span id=\\"2\\">&nbsp;</span><span id=\\"3\\">ر</span><span id=\\"4\\">َ</span><span id=\\"5\\">ج</span><span id=\\"6\\">َ</span><span id=\\"7\\">ع</span><span id=\\"8\\">َ</span><span id=\\"9\\">&nbsp;</span><span id=\\"10\\">ا</span><span id=\\"11\\">ل</span><span id=\\"12\\">ن</span><span id=\\"13\\">ّ</span><span id=\\"14\\">َ</span><span id=\\"15\\">ا</span><span id=\\"16\\">س</span><span id=\\"17\\">ُ</span><span id=\\"18\\">&nbsp;</span><span id=\\"19\\">و</span><span id=\\"20\\">َ</span><span id=\\"21\\">&nbsp;</span><span id=\\"22\\">د</span><span id=\\"23\\">َ</span><span id=\\"24\\">خ</span><span id=\\"25\\">َ</span><span id=\\"26\\">ل</span><span id=\\"27\\">ُ</span><span id=\\"28\\">و</span><span id=\\"29\\">ْ</span><span id=\\"30\\">ا</span><span id=\\"31\\">&nbsp;</span><span id=\\"32\\">ف</span><span id=\\"33\\">ِ</span><span id=\\"34\\">ي</span><span id=\\"35\\">ْ</span><span id=\\"36\\">&nbsp;</span><span id=\\"37\\">ب</span><span id=\\"38\\">َ</span><span id=\\"39\\">ي</span><span id=\\"40\\">ْ</span><span id=\\"41\\">ت</span><span id=\\"42\\">ِ</span><span id=\\"43\\">&nbsp;</span><span id=\\"44\\">ا</span><span id=\\"45\\">ل</span><span id=\\"46\\">ْ</span><span id=\\"47\\">أ</span><span id=\\"48\\">َ</span><span id=\\"49\\">ص</span><span id=\\"50\\">ْ</span><span id=\\"51\\">ن</span><span id=\\"52\\">َ</span><span id=\\"53\\">ا</span><span id=\\"54\\">م</span><span id=\\"55\\">ِ</span></div>
  </div>
  <div dir=\\"rtl\\"><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"اسم إن\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">اسم إن</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"اسم كان\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">اسم كان</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"جارّ\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">جارّ</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"حال\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">حال</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"خبر\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">خبر</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"خبر إن\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">خبر إن</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"خبر كان\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">خبر كان</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"ذو الحال\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">ذو الحال</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"عطف\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">عطف</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"فاعل\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">فاعل</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"فعل\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">فعل</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مفعول به\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مفعول به</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مفعول فيه\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مفعول فيه</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مفعول له\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مفعول له</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مفعول مطلق\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مفعول مطلق</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مجرور\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مجرور</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"معطوف\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">معطوف</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"معطوف عليه\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">معطوف عليه</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مبتدأ\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مبتدأ</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مضاف\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مضاف</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"مضاف إليه\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">مضاف إليه</span></span></button><button type=\\"button\\" class=\\"mx-2 px-2 pb-5 pt-6 primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default\\" id=\\"متعلّق\\"><span class=\\"v-btn__content\\"><span class=\\"arabic\\">متعلّق</span></span></button></div>
</div>"
`;
